marimo/_output/formatters/df_formatters.py

```diff
@@ -206,3 +208,368 @@ def _show_connect_df(
+            is_lazy = not ibis.options.interactive or has_unbound
+
+            return is_lazy, has_unbound
+
+        def _format_expr_and_sql(
+            expr, has_unbound: bool, expr_type: str = "Expression"
+        ) -> tuple[KnownMimeType, str]:
+            """Format expression as tabs with Expression and SQL (if backend supports SQL)."""
+            if has_unbound:
+                expr_content = f"Contains unbound tables - cannot execute\n\n{expr._noninteractive_repr()}"
+                # For unbound expressions, always try SQL generation
+                sql = ibis.to_sql(expr)
+                return tabs.tabs(
+                    {
+                        expr_type: plain_text(expr_content),
+                        "SQL": f"```sql\n{sql!s}\n```",
+                    }
+                )._mime_()
+            else:
+                # For bound expressions, check if backend supports SQL.
+                # Note: We are forced to use this private method instead of __repr__.
+                # Because ibis.options.interactive is set AND the it's unbound the __repr__ will call `to_pyarrow` and crash.
+                # https://github.com/ibis-project/ibis/blob/6b9f80c78e70c9013633ef89e5073a2081c49990/ibis/expr/types/core.py#L53
+                expr_content = expr._noninteractive_repr()
+                backend = expr.get_backend()
+                supports_sql = isinstance(backend, SQLBackend)
+
+                if supports_sql:
+                    sql = ibis.to_sql(expr)
+                    return tabs.tabs(
+                        {
+                            expr_type: plain_text(expr_content),
+                            "SQL": f"```sql\n{sql!s}\n```",
+                        }
+                    )._mime_()
+                else:
+                    # Non-SQL backend, just show expression
+                    return ("text/plain", expr_content)
+
+        def _format_ibis_expression(expr) -> tuple[KnownMimeType, str]:
+            """Format Ibis expressions.
+
+            Interactive mode: Shows data as interactive table
+            Non-interactive or unbound: Shows expression + SQL if it's a SQLBackend
+            """
+            try:
+                is_lazy, has_unbound = _is_lazy_display(expr)
+
+                if is_lazy:
+                    return _format_expr_and_sql(expr, has_unbound)
+                else:
+                    return table(
+                        expr, selection=None, pagination=True
+                    )._mime_()
+
+            except BaseException as e:
+                LOGGER.error("Failed to format Ibis expression: %s", e)
+                # Simple fallback - just show the expression as text
+                try:
+                    return ("text/plain", str(expr))
+                except Exception:
+                    return (
+                        "text/plain",
+                        f"<{type(expr).__name__} object - could not display>",
+                    )
```

Comment:
i think this can be DRYed up:

```
def render_plain_text():
      try:
          return ("text/plain", str(expr)) 1 1Code has comments. Press enter to view.
      except Exception:
          return (
              "text/plain",
              f"<{type(expr).__name__} object - could not display>",
          )
```